helix-term/src/ui/document/text_render.rs

```diff
@@ -0,0 +1,375 @@
+use std::borrow::Cow;
+use std::cmp::min;
+
+use helix_core::graphemes::grapheme_width;
+use helix_core::str_utils::char_to_byte_idx;
+use helix_core::Position;
+use helix_view::editor::{WhitespaceConfig, WhitespaceRenderValue};
+use helix_view::graphics::Rect;
+use helix_view::theme::Style;
+use helix_view::{editor, Document, Theme};
+use tui::buffer::Buffer as Surface;
+
+#[derive(Debug)]
+/// Various constants required for text rendering.
+pub struct TextRenderConfig {
+    pub text_style: Style,
+    pub whitespace_style: Style,
+    pub indent_guide_char: String,
+    pub indent_guide_style: Style,
+    pub newline: String,
+    pub nbsp: String,
+    pub space: String,
+    pub tab: String,
+    pub tab_width: u16,
+    pub starting_indent: usize,
+}
+
+impl TextRenderConfig {
+    pub fn new(
+        doc: &Document,
+        editor_config: &editor::Config,
+        theme: &Theme,
+        offset: &Position,
+    ) -> TextRenderConfig {
+        let WhitespaceConfig {
+            render: ws_render,
+            characters: ws_chars,
+        } = &editor_config.whitespace;
+
+        let tab_width = doc.tab_width();
+        let tab = if ws_render.tab() == WhitespaceRenderValue::All {
+            std::iter::once(ws_chars.tab)
+                .chain(std::iter::repeat(ws_chars.tabpad).take(tab_width - 1))
+                .collect()
+        } else {
+            " ".repeat(tab_width)
+        };
+        let newline = if ws_render.newline() == WhitespaceRenderValue::All {
+            ws_chars.newline.into()
+        } else {
+            " ".to_owned()
+        };
+
+        let space = if ws_render.space() == WhitespaceRenderValue::All {
+            ws_chars.space.into()
+        } else {
+            " ".to_owned()
+        };
+        let nbsp = if ws_render.nbsp() == WhitespaceRenderValue::All {
+            ws_chars.nbsp.into()
+        } else {
+            " ".to_owned()
+        };
+
+        let text_style = theme.get("ui.text");
+
+        TextRenderConfig {
+            indent_guide_char: editor_config.indent_guides.character.into(),
+            newline,
+            nbsp,
+            space,
+            tab_width: tab_width as u16,
+            tab,
+            whitespace_style: theme.get("ui.virtual.whitespace"),
+            starting_indent: (offset.col / tab_width)
+                + editor_config.indent_guides.skip_levels as usize,
+            indent_guide_style: text_style.patch(
+                theme
+                    .try_get("ui.virtual.indent-guide")
+                    .unwrap_or_else(|| theme.get("ui.virtual.whitespace")),
+            ),
+            text_style,
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum Grapheme<'a> {
+    Space,
+    Nbsp,
+    Tab,
+    Other { raw: Cow<'a, str>, width: u16 },
```

Comment:
Is it worth storing the `width` as a field? As `unicode-width` uses a lookup table so it should be O(1).

Reply:
It's a bit more expensive than that, it has a few layers of subtables (0,1,2) + it has to unpack the subtable

Reply:
As @archseer mentioned it's not quite free and the cost of storing the `u16` here is not that large. In fact due to alignment (`Cow` is aligned to `8` bytes) it might actually be free as the size of the enum might not change at all (didn't check tought).  Note that if we cared about the size of this enum enough for that to matter using [beef](https://github.com/maciejhirsz/beef) to get a faster/smaller `cow` would be more worthwhile then saving the `width`. It also doesn't introduce any complexity in the code (it reduces it in fact) so I don't see a downside.